jax/_src/core.py

```diff
@@ -2061,7 +2063,7 @@ class DShapedArray(UnshapedArray):
   array_abstraction_level: int = 3
 
   def __init__(self, shape, dtype, weak_type=False):
-    self.shape = shape
+    self.shape = tuple(d.val if isinstance(d, Literal) else d for d in shape)
```

Comment:
Maybe this should be handled in the caller? Not sure... it just seems like a surprising way to call the constructor.

Reply:
Great question! That was what I originally tried to do, but I struggled to track down all the callers. Let me take another look to see how painful it would be.

Reply:
OK it looks like there's actually only one caller that matters (`DynamicJaxprTrace.newvar`), so I moved this logic there, and added a check that we never see `Literal` objects when constructing a `DShapedArray`. Thanks!